File: frontend/src/types/schema.ts
// src/types/schema.ts
// Types matching backend Pydantic models (app/models/schema.py)

export type SchemaFormat = "AVRO" | "JSON" | "PROTOBUF";

export interface SubjectInfo {
  subject: string;
  format: SchemaFormat;
  latest_version: number;
  version_count: number;
  schema_id: number | null;
  description: string | null;
  owner_team: string | null;
  tags: string[];
}

export interface SchemaDetail {
  subject: string;
  version: number;
  schema_id: number;
  format: SchemaFormat;
  schema_content: Record<string, unknown>;
  references: SchemaReference[];
  registered_at: string | null;
}

export interface SchemaVersion {
  version: number;
  schema_id: number;
  format: SchemaFormat;
  schema_content: Record<string, unknown>;
  registered_at: string | null;
}

export interface SchemaReference {
  name: string;
  subject: string;
  version: number;
}

export type DiffChangeType = "added" | "removed" | "modified" | "unchanged";

export interface FieldDiff {
  field_path: string;
  change_type: DiffChangeType;
  old_value: string | Record<string, unknown> | null;
  new_value: string | Record<string, unknown> | null;
  details: string | null;
}

export interface SchemaDiff {
  subject: string;
  version_from: number;
  version_to: number;
  format: SchemaFormat;
  changes: FieldDiff[];
  is_breaking: boolean;
  summary: string;
}

export interface CompatibilityInfo {
  subject: string;
  compatibility: string;
}

File: backend/app/models/schema.py
"""
event7 - Schema Models
Modèles Pydantic unifiés pour les schemas, versions et diffs.
Ces modèles sont le contrat entre le backend et le frontend.
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class SchemaFormat(str, Enum):
    AVRO = "AVRO"
    JSON_SCHEMA = "JSON"
    PROTOBUF = "PROTOBUF"


class SubjectInfo(BaseModel):
    """Vue résumée d'un subject (pour les listes)"""

    subject: str
    format: SchemaFormat = SchemaFormat.AVRO
    latest_version: int = 1
    version_count: int = 1
    schema_id: int | None = None

    # Enrichissement (optionnel, depuis Supabase)
    description: str | None = None
    owner_team: str | None = None
    tags: list[str] = Field(default_factory=list)
    data_layer: str | None = None

class SchemaDetail(BaseModel):
    """Détail complet d'un schema à une version donnée"""

    subject: str
    version: int
    schema_id: int
    format: SchemaFormat
    schema_content: dict  # Le schema brut (Avro, JSON Schema...)
    references: list["SchemaReference"] = Field(default_factory=list)
    registered_at: datetime | None = None

    # Confluent Data Contracts (optional — only populated by Confluent provider)
    rule_set: dict | None = None  # {"domainRules": [...], "migrationRules": [...]}
    metadata: dict | None = None  # {"properties": {...}, "tags": {...}}


class SchemaVersion(BaseModel):
    """Une version dans l'historique"""

    version: int
    schema_id: int
    format: SchemaFormat
    schema_content: dict
    registered_at: datetime | None = None


class SchemaReference(BaseModel):
    """Référence entre schemas"""

    name: str  # Nom de la référence
    subject: str  # Subject référencé
    version: int  # Version référencée


class DiffChangeType(str, Enum):
    ADDED = "added"
    REMOVED = "removed"
    MODIFIED = "modified"
    UNCHANGED = "unchanged"


class FieldDiff(BaseModel):
    """Diff d'un champ individuel"""

    field_path: str  # Ex: "order.items[].price"
    change_type: DiffChangeType
    old_value: str | dict | None = None
    new_value: str | dict | None = None
    details: str | None = None  # Ex: "type changed from int to long"


class SchemaDiff(BaseModel):
    """Résultat complet d'un diff entre deux versions"""

    subject: str
    version_from: int
    version_to: int
    format: SchemaFormat
    changes: list[FieldDiff] = Field(default_factory=list)
    is_breaking: bool = False
    summary: str = ""  # Ex: "3 fields added, 1 removed, 2 modified"
